#  Module 1

#  Introduction to the problem

Importing libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Importing dataset

## 1.1 

In [3]:
#Loading Dateset
df = pd.read_csv('Datasets/INFY.csv')

In [4]:
#First five rows of dataset
df.head()

Symbol Series         Date  Prev Close  Open Price  High Price  Low Price  \
0   INFY     EQ  15-May-2017      964.25       963.5      963.50     949.10   
1   INFY     EQ  16-May-2017      951.55       953.1      960.15     946.95   
2   INFY     EQ  17-May-2017      955.00       951.6      958.45     943.85   
3   INFY     EQ  18-May-2017      952.80       943.0      973.90     942.85   
4   INFY     EQ  19-May-2017      961.75       961.5      962.70     947.85   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      953.25       951.55         951.49                3648582   
1      956.00       955.00         952.92                3065084   
2      952.90       952.80         949.48                1457754   
3      960.25       961.75         962.61                4028924   
4      957.40       957.95         954.18                2128698   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
0  3.471580e+09          75335          3052819                   83.67  
1  2.920775e+09          71808          1858063                   60.62  
2  1.384110e+09          75429           792251                   54.35  
3  3.878282e+09         120990          2309450                   57.32  
4  2.031155e+09          88897          1457747                   68.48

In [5]:
#Last five rows of dataset
df.tail()

Symbol Series         Date  Prev Close  Open Price  High Price  Low Price  \
491   INFY     EQ  07-May-2019      718.40      720.40      727.90      718.7   
492   INFY     EQ  08-May-2019      724.55      721.00      726.00      716.5   
493   INFY     EQ  09-May-2019      719.35      719.15      730.30      715.8   
494   INFY     EQ  10-May-2019      721.05      720.80      726.00      715.0   
495   INFY     EQ  13-May-2019      716.85      716.55      726.45      712.1   

     Last Price  Close Price  Average Price  Total Traded Quantity  \
491      724.40       724.55         724.51                5747536   
492      718.95       719.35         721.65                6247642   
493      722.00       721.05         724.48                7396525   
494      716.50       716.85         717.88                8352335   
495      718.00       719.70         720.09                6808993   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
491  4.164127e+09         104342          3690425                   64.21  
492  4.508603e+09         162077          3950857                   63.24  
493  5.358635e+09         242225          4808832                   65.01  
494  5.995942e+09         181229          5545377                   66.39  
495  4.903098e+09         201194          4745909                   69.70

In [6]:
#Dimensions of our datset
df.shape

(496, 15)

There are 496 rows and 15 columns in our dataset df

In [7]:
#Looking at brief info of each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 496 entries, 0 to 495
Data columns (total 15 columns):
Symbol                    496 non-null object
Series                    496 non-null object
Date                      496 non-null object
Prev Close                496 non-null float64
Open Price                496 non-null float64
High Price                496 non-null float64
Low Price                 496 non-null float64
Last Price                496 non-null float64
Close Price               496 non-null float64
Average Price             496 non-null float64
Total Traded Quantity     496 non-null int64
Turnover                  496 non-null float64
No. of Trades             496 non-null int64
Deliverable Qty           496 non-null int64
% Dly Qt to Traded Qty    496 non-null float64
dtypes: float64(9), int64(3), object(3)
memory usage: 52.4+ KB


In [8]:
#Number of missing values in each column
df.isnull().sum()

Symbol                    0
Series                    0
Date                      0
Prev Close                0
Open Price                0
High Price                0
Low Price                 0
Last Price                0
Close Price               0
Average Price             0
Total Traded Quantity     0
Turnover                  0
No. of Trades             0
Deliverable Qty           0
% Dly Qt to Traded Qty    0
dtype: int64

There are no missing values in our dataset

In [9]:
#Number of elements in each column
df.count()

Symbol                    496
Series                    496
Date                      496
Prev Close                496
Open Price                496
High Price                496
Low Price                 496
Last Price                496
Close Price               496
Average Price             496
Total Traded Quantity     496
Turnover                  496
No. of Trades             496
Deliverable Qty           496
% Dly Qt to Traded Qty    496
dtype: int64

In [10]:
#Acquiring all stocks of category EQ
df = df[(df['Series'] == 'EQ')]

In [11]:
#After processing dimensions of our dataset
df.shape

(494, 15)

There are 494 stocks of category EQ

In [12]:
#Checking if any row is still left to be removed
(df['Series'] != 'EQ').sum()

0

In [13]:
#Renumbering the index of our dataset
df.reset_index(inplace = True)
df = df.drop(['index'], axis = 1)
df

Symbol Series         Date  Prev Close  Open Price  High Price  Low Price  \
0     INFY     EQ  15-May-2017      964.25      963.50      963.50     949.10   
1     INFY     EQ  16-May-2017      951.55      953.10      960.15     946.95   
2     INFY     EQ  17-May-2017      955.00      951.60      958.45     943.85   
3     INFY     EQ  18-May-2017      952.80      943.00      973.90     942.85   
4     INFY     EQ  19-May-2017      961.75      961.50      962.70     947.85   
..     ...    ...          ...         ...         ...         ...        ...   
489   INFY     EQ  07-May-2019      718.40      720.40      727.90     718.70   
490   INFY     EQ  08-May-2019      724.55      721.00      726.00     716.50   
491   INFY     EQ  09-May-2019      719.35      719.15      730.30     715.80   
492   INFY     EQ  10-May-2019      721.05      720.80      726.00     715.00   
493   INFY     EQ  13-May-2019      716.85      716.55      726.45     712.10   

     Last Price  Close Price  Average Price  Total Traded Quantity  \
0        953.25       951.55         951.49                3648582   
1        956.00       955.00         952.92                3065084   
2        952.90       952.80         949.48                1457754   
3        960.25       961.75         962.61                4028924   
4        957.40       957.95         954.18                2128698   
..          ...          ...            ...                    ...   
489      724.40       724.55         724.51                5747536   
490      718.95       719.35         721.65                6247642   
491      722.00       721.05         724.48                7396525   
492      716.50       716.85         717.88                8352335   
493      718.00       719.70         720.09                6808993   

         Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  
0    3.471580e+09          75335          3052819                   83.67  
1    2.920775e+09          71808          1858063                   60.62  
2    1.384110e+09          75429           792251                   54.35  
3    3.878282e+09         120990          2309450                   57.32  
4    2.031155e+09          88897          1457747                   68.48  
..            ...            ...              ...                     ...  
489  4.164127e+09         104342          3690425                   64.21  
490  4.508603e+09         162077          3950857                   63.24  
491  5.358635e+09         242225          4808832                   65.01  
492  5.995942e+09         181229          5545377                   66.39  
493  4.903098e+09         201194          4745909                   69.70  

[494 rows x 15 columns]

In [14]:
#Datatype of each column
df.dtypes

Symbol                     object
Series                     object
Date                       object
Prev Close                float64
Open Price                float64
High Price                float64
Low Price                 float64
Last Price                float64
Close Price               float64
Average Price             float64
Total Traded Quantity       int64
Turnover                  float64
No. of Trades               int64
Deliverable Qty             int64
% Dly Qt to Traded Qty    float64
dtype: object

## 1.2 

Maximum price for last 90 days

In [15]:
df.iloc[-90:-1]['Close Price'].max()

767.25

Minimum price for last 90 days

In [16]:
df.iloc[-90:-1]['Close Price'].min()

658.95

Average price for last 90 days

In [17]:
df.iloc[-90:-1]['Close Price'].mean()

728.702808988764

## 1.3

In [18]:
#Converting datatype of Date column
df['Date'] = df['Date'].astype('datetime64')

In [19]:
#Looking if the datatype has got changed
df.dtypes

Symbol                            object
Series                            object
Date                      datetime64[ns]
Prev Close                       float64
Open Price                       float64
High Price                       float64
Low Price                        float64
Last Price                       float64
Close Price                      float64
Average Price                    float64
Total Traded Quantity              int64
Turnover                         float64
No. of Trades                      int64
Deliverable Qty                    int64
% Dly Qt to Traded Qty           float64
dtype: object

In [20]:
#Range of dates in Date column
(df['Date'].max() - df['Date'].min())

Timedelta('728 days 00:00:00')

## 1.4 

In [21]:
#Creating new column Month
df['Month'] = df['Date'].dt.month

In [22]:
#Creating new column Year
df['Year'] = df['Date'].dt.year

Creating new array VWAP containing volume weighted average price grouped by month

In [23]:
#Creating new dataframe df_VWAP for further processing
df_VWAP = df

In [24]:
#Creating new column Price * Volume
df_VWAP['Price * Volume'] = df['Close Price'] * df['Total Traded Quantity']

In [24]:
#First five rows of our dataframe
df_VWAP.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0   INFY     EQ 2017-05-15      964.25       963.5      963.50     949.10   
1   INFY     EQ 2017-05-16      951.55       953.1      960.15     946.95   
2   INFY     EQ 2017-05-17      955.00       951.6      958.45     943.85   
3   INFY     EQ 2017-05-18      952.80       943.0      973.90     942.85   
4   INFY     EQ 2017-05-19      961.75       961.5      962.70     947.85   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      953.25       951.55         951.49                3648582   
1      956.00       955.00         952.92                3065084   
2      952.90       952.80         949.48                1457754   
3      960.25       961.75         962.61                4028924   
4      957.40       957.95         954.18                2128698   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  3.471580e+09          75335          3052819                   83.67   
1  2.920775e+09          71808          1858063                   60.62   
2  1.384110e+09          75429           792251                   54.35   
3  3.878282e+09         120990          2309450                   57.32   
4  2.031155e+09          88897          1457747                   68.48   

   Month  Year  Price * Volume  
0      5  2017    3.471808e+09  
1      5  2017    2.927155e+09  
2      5  2017    1.388948e+09  
3      5  2017    3.874818e+09  
4      5  2017    2.039186e+09

In [25]:
#Grouping of our datframe on Month and year and calculating sum of Total Traded Quantity and Price * Volume Column
df_VWAP_group = df_VWAP.groupby(['Month', 'Year'])[['Total Traded Quantity', 'Price * Volume']].sum()

VWAP(Volume Weighted Average Price) = sum(price * volume) / sum(volume)

In [26]:
#Calculating VWAP
VWAP = df_VWAP_group['Price * Volume'] / df_VWAP_group['Total Traded Quantity']

In [27]:
#Flattening up the array
VWAP = VWAP.reset_index()

In [28]:
VWAP.rename(columns = {0:'VWAP'}, inplace = True)
VWAP.sort_values(by = ['Year', 'Month'], inplace = True)
VWAP

Month  Year         VWAP
8       5  2017   969.171663
11      6  2017   948.202546
13      7  2017   979.204067
15      8  2017   919.396317
17      9  2017   897.396520
19     10  2017   931.858118
21     11  2017   969.989288
23     12  2017  1011.833961
0       1  2018  1114.434642
2       2  2018  1137.523710
4       3  2018  1162.184207
6       4  2018  1153.567462
9       5  2018  1197.926656
12      6  2018  1263.282758
14      7  2018  1329.247518
16      8  2018  1396.527179
18      9  2018   752.570451
20     10  2018   688.817051
22     11  2018   650.652805
24     12  2018   670.195881
1       1  2019   708.054850
3       2  2019   746.052269
5       3  2019   724.365636
7       4  2019   739.025955
10      5  2019   721.403986

## 1.5 

Function to calculate average price for last N days

In [29]:
def average_price(df, N):
    return df.iloc[-N:-1]['Close Price'].mean()

Function to calculate profit/loss percentage for last N days

In [30]:
def profit_loss_perc(df, N):
    p = (df.iloc[-1]['Close Price'] - df.iloc[-N]['Close Price']) / df.iloc[-N]['Close Price']
    p = p * 100
    return p

In [31]:
#Calculating Average Prices and Profit/ Loss percentages

#Last 1 week
print("Average Price for last 1 week")
print("{:.2f}".format(average_price(df, 7)))
print("Profit/Loss percentage for last 1 week")
print("{:.2f}".format(profit_loss_perc(df, 7)))

#Last 2 weeks
print("Average Price for last 2 weeks")
print("{:.2f}".format(average_price(df, 14)))
print("Profit/Loss percentage for last 2 weeks")
print("{:.2f}".format(profit_loss_perc(df, 14)))

#Last 1 month
print("Average Price for last 1 month")
print("{:.2f}".format(average_price(df, 30)))
print("Profit/Loss percentage for last 1 month")
print("{:.2f}".format(profit_loss_perc(df, 30)))

#Last 3 months
print("Average Price for last 3 months")
print("{:.2f}".format(average_price(df, 90)))
print("Profit/Loss percentage for last 3 months")
print("{:.2f}".format(profit_loss_perc(df, 90)))

#Last 6 months
print("Average Price for last 6 months")
print("{:.2f}".format(average_price(df, 180)))
print("Profit/Loss percentage for last 6 months")
print("{:.2f}".format(profit_loss_perc(df, 180)))

#Last 1 year
print("Average Price for last 1 year")
print("{:.2f}".format(average_price(df, 365)))
print("Profit/Loss percentage for last 1 year")
print("{:.2f}".format(profit_loss_perc(df, 365)))

Average Price for last 1 week
720.63
Profit/Loss percentage for last 1 week
-0.54
Average Price for last 2 weeks
727.55
Profit/Loss percentage for last 2 weeks
-0.19
Average Price for last 1 month
737.39
Profit/Loss percentage for last 1 month
-1.37
Average Price for last 3 months
728.70
Profit/Loss percentage for last 3 months
9.22
Average Price for last 6 months
758.91
Profit/Loss percentage for last 6 months
-48.90
Average Price for last 1 year
967.56
Profit/Loss percentage for last 1 year
-25.88


## 1.6 

In [32]:
#Creating new column Day_Perc_Change
df['Day_Perc_Change'] = df['Close Price'].pct_change() * 100
df['Day_Perc_Change'].head()

0         NaN
1    0.362566
2   -0.230366
3    0.939337
4   -0.395113
Name: Day_Perc_Change, dtype: float64

In [33]:
#Replacing missing values with 0
df['Day_Perc_Change'].replace(np.nan, 0, inplace = True)

In [34]:
#First five values
df['Day_Perc_Change'].head()

0    0.000000
1    0.362566
2   -0.230366
3    0.939337
4   -0.395113
Name: Day_Perc_Change, dtype: float64

## 1.7

Function trend to apply on 'Day_Perc_Change' column

In [35]:
def trend(x):
    if x >= -0.5 and x <= 0.5 :
        a = 'Slight or No change'
    elif x >= 0.5 and x <= 1:
        a = 'Slight positive'
    elif x >= -1 and x <= -0.5:
        a = 'Slight negative'
    elif x >= 1 and x <= 3:
        a = 'Positive'
    elif x >= -3 and x <= -1:
        a = 'Negative'
    elif x >= 3 and x <= 7:
        a = 'Among top gainers'
    elif x >= -7 and x <= -3:
        a = 'Among top losers'
    elif x > 7:
        a = 'Bull run'
    elif x < -7:
        a = 'Bear drop'
        
    return a

In [36]:
#Creating new column Trend
df['Trend'] = df['Day_Perc_Change'].apply(trend)

In [37]:
#First five rows of our dataframe
df.head()

Symbol Series       Date  Prev Close  Open Price  High Price  Low Price  \
0   INFY     EQ 2017-05-15      964.25       963.5      963.50     949.10   
1   INFY     EQ 2017-05-16      951.55       953.1      960.15     946.95   
2   INFY     EQ 2017-05-17      955.00       951.6      958.45     943.85   
3   INFY     EQ 2017-05-18      952.80       943.0      973.90     942.85   
4   INFY     EQ 2017-05-19      961.75       961.5      962.70     947.85   

   Last Price  Close Price  Average Price  Total Traded Quantity  \
0      953.25       951.55         951.49                3648582   
1      956.00       955.00         952.92                3065084   
2      952.90       952.80         949.48                1457754   
3      960.25       961.75         962.61                4028924   
4      957.40       957.95         954.18                2128698   

       Turnover  No. of Trades  Deliverable Qty  % Dly Qt to Traded Qty  \
0  3.471580e+09          75335          3052819                   83.67   
1  2.920775e+09          71808          1858063                   60.62   
2  1.384110e+09          75429           792251                   54.35   
3  3.878282e+09         120990          2309450                   57.32   
4  2.031155e+09          88897          1457747                   68.48   

   Month  Year  Price * Volume  Day_Perc_Change                Trend  
0      5  2017    3.471808e+09         0.000000  Slight or No change  
1      5  2017    2.927155e+09         0.362566  Slight or No change  
2      5  2017    1.388948e+09        -0.230366  Slight or No change  
3      5  2017    3.874818e+09         0.939337      Slight positive  
4      5  2017    2.039186e+09        -0.395113  Slight or No change

## 1.8 

In [38]:
#Grouping dataframe on Trend
df_trend_group = df.groupby('Trend')[['Total Traded Quantity']]

In [39]:
#Average Total Traded Quantity
df_trend_group.mean()

Total Traded Quantity
Trend                                     
Among top gainers             1.259063e+07
Among top losers              1.463854e+07
Bear drop                     4.878630e+07
Negative                      5.634151e+06
Positive                      6.655530e+06
Slight negative               5.454672e+06
Slight or No change           4.647983e+06
Slight positive               5.848807e+06

In [40]:
#Median Total Traded Quantity
df_trend_group.median()

Total Traded Quantity
Trend                                     
Among top gainers               12042170.5
Among top losers                10985897.0
Bear drop                       48786302.0
Negative                         5088554.0
Positive                         5523811.5
Slight negative                  4616803.0
Slight or No change              4043539.5
Slight positive                  5301784.0

## 1.9

In [41]:
#Saving dataframe
df.to_csv('Saved Datasets/week2.csv')